mooncake-store/src/hf3fs/threefs_file.cpp

```diff
@@ -0,0 +1,374 @@
+#include "file_interface.h"
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/file.h>
+#include <cstring>
+#include <stdexcept>
+
+namespace mooncake {
+
+ThreeFSFile::ThreeFSFile(const std::string& filename, int fd, USRBIOResourceManager* resource_manager) 
+    : StorageFile(filename, fd), resource_manager_(resource_manager) {}
+
+ThreeFSFile::~ThreeFSFile() {
+    // Deregister and close file descriptor
+    if (fd_ >= 0) {
+        hf3fs_dereg_fd(fd_);
+        if (close(fd_) == -1) {
+            LOG(WARNING) << "Failed to close file: " << filename_;
+        }
+        fd_ = -1;
+    }
+
+    // Delete potentially corrupted file if write failed
+    if (error_code_ == ErrorCode::FILE_WRITE_FAIL) {
+        if (::unlink(filename_.c_str()) == -1) {
+            LOG(ERROR) << "Failed to delete corrupted file: " << filename_;
+        } else {
+            LOG(INFO) << "Deleted corrupted file: " << filename_;
+        }
+    }
+}
+
+ssize_t ThreeFSFile::write(const std::string& buffer, size_t length) {
```

Comment:
Perchance, eschew std::string as the buffer argument? Consider adopting a span or a plain char* instead.